### PR TITLE
[CUDA] Optimize QMoE SoftmaxTopK router for small-batch decode

### PR DESCRIPTION
**Pull request by tianleiwu**
### Description

Optimizes the CUDA QMoE router top-k (`LaunchSoftmaxTopK`) for small-batch / autoregressive decode by replacing the old one-thread-per-row hot path with parallel CUB and warp-level top-k kernels. The dispatch now uses the fastest specialized path for common MoE expert counts while preserving the existing softmax normalization and deterministic lower-index tie-breaking semantics.

This PR also factors the warp-level top-k sorting code into a reusable CUDA helper header and adds direct CUDA-internal tests so the new routing paths are covered independently of higher-level QMoE tests.

### Motivation and Context

The previous router path launched a 256-thread block per row but did all top-k work in a single thread. In decode scenarios such as `num_rows == 1`, that made the router latency-bound on a serial scan of all expert logits and turned `SoftmaxTopKKernel` into a major MoE decode bottleneck.

For a Qwen3-style MoE workload with 256 experts, top-8 routing, and 40 MoE layers, the original router accounted for roughly 50% of decode GPU time. Moving the work to block/warp-parallel kernels removes that bottleneck while keeping the same output ordering and scaling behavior.

### Key Changes

| Area | Change |
|---|---|
| QMoE router dispatch | Adds `DispatchSoftmaxTopK` routing for `k <= 64` and `num_experts <= 1024`, with a fallback to the original scalar kernel for larger or uncommon shapes. |
| Tiny expert counts | Adds `SoftmaxTopKWarpBitonicKernel` for `num_experts <= 32`, using one warp per row and in-register bitonic sorting via warp shuffles. |
| Small expert counts | Adds `SoftmaxTopKWarpMergeKernel` for `32 < num_experts <= 64`, using a single warp and CUB warp merge sort. |
| Larger common MoE counts | Uses `SoftmaxTopKMergeKernel` with CUB block merge sort for `num_experts <= 128`, `256`, `512`, and `1024`. |
| Reusable top-k helpers | Adds `onnxruntime/core/providers/cuda/cu_inc/topk_warp_sort.cuh` with reusable warp bitonic and warp merge sort helpers. |
| Stable tie-breaking | Packs `(score, index)` into a `uint64_t` stable sort key for the CUB merge paths, matching onnxruntime-genai's lower-index tie-breaking and avoiding compound comparators. |
| Softmax cleanup | Factors shared softmax scale, safe reciprocal, top-k normalization, warp reduction, and CUB block reduction helpers to keep the optimized kernels consistent. |
| Tests | Adds CUDA-internal `SoftmaxTopK_*` tests covering warp bitonic, warp merge, block merge, stable ties, normalization, `float`, `half`, and `bfloat16`. |

### Performance

H200 measurements for the target QMoE decode scenario showed the router cost dropping from roughly `5.56 ms/token` to `0.17 ms/token`, improving end-to-end Qwen3.6-35B-A3B INT4 decode throughput from about `80 tok/s` to `113 tok/s`.

Additional profiling of the `32 < num_experts <= 64` warp merge path showed the packed `uint64_t` stable sort key is consistently faster than a `{float, int}` struct comparator on H200:

| Experts | Sort-only packed/struct | Full softmax+top-k packed/struct |
|---:|---:|---:|
| 33 | 0.680x | 0.704x |
| 48 | 0.672x | 0.695x |
| 64 | 0.673x | 0.696x |

### Testing

- `lintrunner -a`
- `ninja onnxruntime_providers_cuda_ut`
- `ninja onnxruntime_provider_test`
- `GTEST_FILTER='CUDA_EP_Unittest.SoftmaxTopK_*' ./onnxruntime_provider_test --gtest_filter='CUDA_EP_Unittest.All'`
- `onnxruntime/test/python/transformers/test_qmoe_cuda.py -k parity` (`44 passed`)

### CODE REVIEW

File: onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu [hunk 1]
```diff
@@ -88,6 +88,255 @@ __global__ void SoftmaxTopKKernel(const T* logits, float* topk_scales, int* topk
...
+  const float inv_sum = (sum_exp > 0.0f) ? (1.0f / sum_exp) : 0.0f;
+  __syncthreads();
+
+  // 3. Top-k via k rounds of block-wide argmax (k is small). Selecting on the
+  // logit value is equivalent to selecting on the softmax probability since the
+  // softmax is monotonic; the probability is recovered only for the winners.
+  for (int r = 0; r < k; ++r) {
+    float best_val = -FLT_MAX;
+    int best_idx = num_experts;  // larger than any valid index -> lowest-index tie-break
+    for (int i = tid; i < num_experts; i += kBlockSize) {
```
Comment (apsonawane) [inline onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu]: This has 3 for loops, can it be optimized?

File: onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu [hunk 2]
```diff
@@ -88,6 +88,255 @@ __global__ void SoftmaxTopKKernel(const T* logits, float* topk_scales, int* topk
...
+  for (int j = 0; j < kItemsPerThread; ++j) local_max = fmaxf(local_max, keys[j]);
+#if CUDART_VERSION >= 12090
+  float block_max = BlockReduce(temp.reduce).Reduce(local_max, ::cuda::maximum<float>());
+#else
+  float block_max = BlockReduce(temp.reduce).Reduce(local_max, cub::Max());
+#endif
+  if (tid == 0) s_max = block_max;
+  __syncthreads();
+  const float max_val = s_max;
+  __syncthreads();
```
Comment (apsonawane) [inline onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu]: Do we need this syncthreads?

File: onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu [hunk 3]
```diff
@@ -88,6 +88,255 @@ __global__ void SoftmaxTopKKernel(const T* logits, float* topk_scales, int* topk
...
+  const float inv_sum = (s_sum > 0.0f) ? (1.0f / s_sum) : 0.0f;
+  __syncthreads();
+
+  // Sort (logit, index) pairs descending. Result stays in a blocked layout, so
+  // sorted rank r lives in thread (r / ipt), item (r % ipt).
+  struct DescendingOp {
+    __device__ bool operator()(const float& a, const float& b) const { return a > b; }
+  };
+  BlockMergeSort(temp.merge).Sort(keys, vals, DescendingOp());
+  __syncthreads();
```
Comment (apsonawane) [inline onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu]: Same here, is syncthreads needed?

[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu]: Good catch — `cub::BlockMergeSort` is not a stable sort, so the comment was wrong and exact-tie routing could be nondeterministic. Fixed in 939da79: the kernel now sorts a composite `LogitIndexKey{logit, index}` with a comparator that orders by descending logit and breaks ties on the lower expert index, so the tie-break no longer relies on stability. Updated the comment accordingly.

File: onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu [hunk 4]
```diff
@@ -88,6 +88,255 @@ __global__ void SoftmaxTopKKernel(const T* logits, float* topk_scales, int* topk
...
+  // Softmax denominator over all experts (needed when normalize_scales is false;
+  // when true it cancels in the top-k normalization but is still correct).
+  float local_max = -FLT_MAX;
+#pragma unroll
+  for (int j = 0; j < kItemsPerThread; ++j) local_max = fmaxf(local_max, keys[j]);
+#if CUDART_VERSION >= 12090
+  float block_max = BlockReduce(temp.reduce).Reduce(local_max, ::cuda::maximum<float>());
+#else
+  float block_max = BlockReduce(temp.reduce).Reduce(local_max, cub::Max());
+#endif
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu]: Aligned to the non-templated `::cuda::maximum()` in 939da79 to match the other CUDA softmax kernels in the repo (e.g. `bert/attention_softmax.cu`, `bert/unfused_attention.cu`).

File: onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu [hunk 5]
```diff
@@ -88,6 +88,255 @@ __global__ void SoftmaxTopKKernel(const T* logits, float* topk_scales, int* topk
...
+#pragma unroll
+  for (int j = 0; j < kItemsPerThread; ++j) {
+    const int idx = tid * kItemsPerThread + j;
+    if (idx < num_experts) local_sum += expf(keys[j] - max_val);
+  }
+#if CUDART_VERSION >= 12090
+  float block_sum = BlockReduce(temp.reduce).Reduce(local_sum, ::cuda::std::plus<float>());
+#else
+  float block_sum = BlockReduce(temp.reduce).Reduce(local_sum, cub::Sum());
+#endif
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu]: Same fix — changed to `::cuda::std::plus()` in 939da79 to match the existing repo usage.

File: onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu (hunk 1 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu]: This is the `SoftmaxTopKKernelBlock` fallback, which only runs for `num_experts > 1024`; the common decode path (and any realistic MoE expert count) goes through the CUB merge-sort kernel above. The three loops are: the outer `k` rounds, the middle coalesced expert scan, and the inner O(r) "already chosen" check (r < k, so just a few iterations). For the rare >1024-expert case with small `k` this is negligible, so I kept it as the simple/correct cold-path fallback rather than optimizing a path that is not hit in practice.

File: onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu (hunk 2 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu]: Good catch — the second barrier was redundant. The barrier right after `s_max` is written already does both jobs: it publishes `s_max` to all threads and separates the two `BlockReduce` calls that share `temp.reduce`. Removed the extra `__syncthreads()` in 939da79 (this matches the single-barrier pattern in `bert/attention_softmax.cu`).

File: onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu (hunk 3 above)
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu]: Removed in 939da79. `Sort()` returns the sorted keys in each threads registers and `temp.merge` is not reused afterwards, so no barrier is needed here; the shared `s_topk` writes that follow are published by the existing barrier just before the `tid == 0` normalization. (I also dropped the analogous redundant barrier after `inv_sum` for the same reason.)

File: onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu [hunk 6]
```diff
@@ -80,14 +131,258 @@ __global__ void SoftmaxTopKKernel(const T* logits, float* topk_scales, int* topk
...
+                         cudaStream_t stream) {
+  // Block-per-row CUB merge sort is the fastest path for the common decode case
+  // (one block fully sorts a row). Pick the smallest capacity that covers
+  // num_experts. k must fit in s_topk (<= 64).
+  const dim3 grid(static_cast<unsigned>(num_rows));
+  if (k <= 64 && num_experts <= 1024) {
+    // Tiny expert counts: a single warp sorts a row entirely in registers via
+    // an in-register bitonic sort (no shared memory). Multiple warps per block
+    // process multiple rows for better occupancy.
+    if (num_experts <= onnxruntime::cuda::topk::kWarpBitonicMaxSize) {
```
[the remark this answers is not in the document]
Comment (tianleiwu) [reply onnxruntime/contrib_ops/cuda/moe/qmoe_kernels.cu:356]: Maybe `ORT_ENFORCE(k <= num_experts);`. I guess such check shall be done earlier, and not need in kernel code.

**Review by apsonawane — APPROVED** [review]
